[WIP][Training] Flux Control LoRA training script

sayakpaul opened the pull request:
Will be updated later. 

This PR has both LoRA and a non-LoRA version for training Control Flux. 

~Number of changed files are more than expected because the `flux-control-lora` branch in merged into this PR branch.~

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_10130). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

a-r-r-o-w commented:
@sayakpaul I think some things like the model card are yet to be updated, and there's probably more things on your mind. Once you mark this ready for review, I can take a deeper look - everything looks great already though! Thanks for involving me in the discussions and looking forward to starting my own training run soon!

sayakpaul commented:
> I think some things like the model card are yet to be updated

I think I have the model cards already updated. But let me know if you think some things should be added/updated. 

Marking this PR as ready to review.

sayakpaul commented:
@apolinario @linoytsaban do you think it could make sense to show users how to _derive_ a LoRA from the full Control fine-tune like how the BFL folks did?

a-r-r-o-w commented on examples/control-lora/README.md:
```diff
@@ -0,0 +1,117 @@
+# Training Control LoRA with Flux
+
+This (experimental) example shows how train Control LoRA with [Flux](https://huggingface.co/black-forest-labs/FLUX.1-dev) to condition it with additional structural controls (like depth maps, poses, etc.). 
```
```suggestion
This (experimental) example shows how to train Control LoRAs with [Flux](https://huggingface.co/black-forest-labs/FLUX.1-dev) by conditioning it with additional structural controls (like depth maps, poses, etc.).
```

a-r-r-o-w commented on examples/control-lora/README.md:
```diff
@@ -0,0 +1,117 @@
+
+This (experimental) example shows how train Control LoRA with [Flux](https://huggingface.co/black-forest-labs/FLUX.1-dev) to condition it with additional structural controls (like depth maps, poses, etc.). 
+
+We simply expand the input channels of Flux.1 Dev from 64 to 128 to allow for additional inputs and then train a regular LoRA on top of it. To account for the newly added input channels, we additional append a LoRA on the underlying layer (`x_embedder`). Inference, however, is performed with the `FluxControlPipeline`. 
```
```suggestion
To incorporate additional condition latents, we expand the input features of Flux.1-Dev from 64 to 128. The first 64 channels correspond to the original input latents to be denoised, while the latter 64 channels correspond to control latents. This expansion happens on the `x_embedder` layer, where the combined latents are projected to the expected feature dimension of rest of the network. Inference is performed using the `FluxControlPipeline`.
```

a-r-r-o-w commented on examples/control-lora/train_control_flux.py:
```diff
@@ -0,0 +1,1202 @@
+    return pixel_latents.to(weight_dtype)
+
+
+def log_validation(flux_transformer, args, accelerator, weight_dtype, step, is_final_validation=False):
```
Very clean!

a-r-r-o-w commented on examples/control-lora/train_control_flux.py:
```diff
@@ -0,0 +1,1202 @@
+# controlnet-{repo_id}
+
+These are Control weights trained on {base_model} with new type of conditioning.
+{img_str}
```
Oh, this is what I missed earlier when I though model card was incomplete. The validation images will be displayed, awesome!

a-r-r-o-w commented on examples/control-lora/train_control_flux.py:
```diff
@@ -0,0 +1,1202 @@
+            img_str += f"![images_{i})](./images_{i}.png)\n"
+
+    model_description = f"""
+# controlnet-{repo_id}
```
```suggestion
# control-lora-{repo_id}
```

or something similar maybe

a-r-r-o-w commented on examples/control-lora/train_control_flux.py:
```diff
@@ -0,0 +1,1202 @@
+                sigmas = get_sigmas(timesteps, n_dim=pixel_latents.ndim, dtype=pixel_latents.dtype)
+                noisy_model_input = (1.0 - sigmas) * pixel_latents + sigmas * noise
+                # Concatenate across channels.
+                # Question: Should we concatenate before adding noise?
```
I think what you have currently is correct. Noisy latent + clean control latent

a-r-r-o-w commented on examples/control-lora/train_control_lora_flux.py:
```diff
@@ -0,0 +1,1353 @@
+        "flux-diffusers",
+        "text-to-image",
+        "diffusers",
+        "control-lora",
```
I think also maybe add `lora` here to make it searchable on https://huggingface.co/models?other=lora as well. No problem if not

sayakpaul replied:
Done.

a-r-r-o-w commented on examples/control-lora/train_control_lora_flux.py:
```diff
@@ -0,0 +1,1353 @@
+        ),
+    )
+    parser.add_argument(
+        "--set_grads_to_none",
```
The default behaviour of pytorch is that grads will be set to None. With this, we default it to False unless specified. Is this intended? I think we should default to pytorch behaviour, and if an option is to be provided, it should be something like `dont_set_grads_to_none`. Not a blocker

sayakpaul replied:
You're right. I will remove it altogether.

a-r-r-o-w commented on examples/control-lora/train_control_lora_flux.py:
```diff
@@ -0,0 +1,1353 @@
+            " Please use `huggingface-cli login` to authenticate with the Hub."
+        )
+    if args.use_lora_bias and args.gaussian_init_lora:
+        raise ValueError("`gaussian` LoRA init scheme isn't supported when `use_lora_bias` is True.")
```
Just wondering why this is the case. We could set initialization to gaussian before the lora bias changes IIRC, no?

sayakpaul replied:
So, we do:

```py
LoraConfig(lora_bias=True, init_lora_weights="gaussian", ...)
```

it errors out. Maybe a question for @BenjaminBossan.

BenjaminBossan replied:
As [discussed earlier](https://github.com/huggingface/peft/pull/2237#discussion_r1860357307), it just hasn't been implemented yet. The question is: If a user wants Gaussian init for the LoRA weights, how do we initialize the LoRA bias? If we have an answer to that, we can enable Gaussian init for `lora_bias`.

a-r-r-o-w reviewed (approved):
Awesome, LGTM! Just some minor comments

sayakpaul commented on examples/control-lora/train_control_lora_flux.py:
```diff
@@ -0,0 +1,1353 @@
+        "text-to-image",
+        "diffusers",
+        "control-lora",
+        "diffusers-training",
```
```suggestion
        "diffusers-training",
        "lora",
```

sayakpaul commented:
@a-r-r-o-w I have addressed your comments, thank you!

If you could review the changes and the README mods (made in https://github.com/huggingface/diffusers/pull/10130/commits/02016caf921e269cad4240b4b193f3831120bb65), that would be helpful!

a-r-r-o-w reviewed (approved):
Let's get the GPUs burning! LoRA testing scripts can be added in a follow-up PR if required but not very important

sayakpaul commented:
Thanks, I think also okay to not add tests just yet. Based on the usage, we can always revisit.

Adenialzz commented:
Hi, thanks for your implementation.

I'm curious why we set model offloading as default and don't have the option to turn it off. In my testing, this can significantly affect training speed ( ~5x). I recommend making model offloading optional in args

sayakpaul commented:
@Adenialzz, thanks for your comment. Feel free to open a PR :)

Adenialzz commented:
Sure. Here https://github.com/huggingface/diffusers/pull/10225 .
